keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,956 @@
+achieved failure) and (2) failures are more easily auto-reverted by upgrade
+orchestration as we are taking smaller and more incremental steps forward,
+which means there is less to “undo” on a failure condition.
+
+In beta, we intend to introduce support for greater compatibility version skew
+(specifically, N-3) so that it would be possible to skip-level upgrade a 
+Kubernetes control-plane, by means of:
+
+- (starting point) binary-version 1.28 (compat-version 1.28)
+- upgrade binary-version to 1.31 (compat-version stays at 1.28 - this is our skip-level binary upgrade)
+- keep binary-version 1.31 while upgrading compat-version to 1.29 (stepwise upgrade of compatibility)
+- keep binary-version 1.31 while upgrading compat-version to 1.30 (stepwise upgrade of compatibility)
+- keep binary-version 1.31 while upgrading compat-version to 1.31 (stepwise upgrade of compatibility)
+
+
+### Goals
+
+- Introduce the metadata necessary to configure features/APIs/storage-versions/CEL
+  features to match the behavior of an older Kubernetes release version
+- A Kubernetes binary with compatibility version set to N, will pass the
+  conformance and e2e tests from Kubernetes release version N.
+- A Kubernetes binary with compatibility version set to N does not enable any
+  changes (storage versions, CEL feature, features) that would prevent it
+  from being rolled back to N-1.
+- The most recent Kubernetes version supports compatibility version being set to
+  the full range of supported versions (N..N-3).
+
+### Non-Goals
+
+- Changes to Cluster API/kubeadm/KIND/minikube to absorb the compatibility versions
+  will be addressed separate from this KEP
+
+## Proposal
+
+### Component Flags
+
+Kubernetes components (apiservers, controller managers, schedulers) will offer a
+`--compatibility-version` flag that can be set to any of the previous three
+minor versions. If unset, the compatibility version defaults to the `<major.minor>`
+version of the binary version.
+
+### Feature Compatibility Versioning
+
+Features will be versioned, i.e.:
+
+```go
+type FeatureSpec struct {
+  //...
+
+  // Version indicates the version this feature spec was introduced.
+	Version semver.Version
+}
+```
+
+When a component starts, feature gates will be compared against the
+compatibility version to determine which features to enable to match the set of
+features that where enabled for the Kubernetes version the compatibility version
+is set to.
+
+Also, `--feature-gates` must behave the same as it did for the Kubernetes
+version the compatibility version is set to. I.e. it must be possible to use
+`--feature-gates` to disable features that were beta, and enable feature that
+were alpha in the Kubernetes version the compatibility version is set to. One
+important implication of this requirement is that feature gating must be kept in
+the Kubenetes codebase until a feature has reached GA (or been removed) for N-3
```

Comment:
Calling this out as critical to be sure that all sigs are aware of.  This means that code must be gate-able far beyond the current lifespan.

Reply:
I find useful to put practical examples to make it easier to understand the implications